src/babel/transformation/transformers/es6/classes.js

```diff
@@ -170,15 +170,31 @@ class ClassTransformer {
 
     this.buildBody();
 
+    var decorators = this.node.decorators;
+    var classCheckRef = classRef;
+
+    if (decorators) {
+      classCheckRef = this.scope.generateUidIdentifier(classRef);
+    }
+
     constructorBody.body.unshift(t.expressionStatement(t.callExpression(file.addHelper("class-call-check"), [
       t.thisExpression(),
-      classRef
+      classCheckRef
     ])));
 
     //
 
-    var decorators = this.node.decorators;
     if (decorators) {
+      if (this.className) {
+        body.unshift(t.variableDeclaration("var", [
+          t.variableDeclarator(classCheckRef)
+        ]));
+
+        body.push(t.expressionStatement(
```

Comment:
Any reason this can't just be a variable declaration instead of splitting it up?